### PR TITLE
feat(noise-suppression): Add noise suppression effect.

### PR DESCRIPTION
**Pull request by andrei-gavrilescu**
Depends on: https://github.com/jitsi/lib-jitsi-meet/pull/2008

This PR creates a noise suppression effect that can be applied to an audio JitsiLocalTrack.
The effect essentially passes the audio MediaStream to a NoiseSuppressior instance, which uses rnnoise to denoise audio frames.
It can be toggled by using the Noise Suppression button.

### CODE REVIEW

**Comment by saghul** [comment]
I'll do a full review tomorrow but quick question: why did you put part of it in LJM? AFAIS it still needs the rnnoise processor to be passed from JM so I guess we could have it all in the app layer?

To put it another way, what benefit does the separation add to LJM users?

**Comment by andrei-gavrilescu** [comment]
> I'll do a full review tomorrow but quick question: why did you put part of it in LJM? AFAIS it still needs the rnnoise processor to be passed from JM so I guess we could have it all in the app layer?
> 
> To put it another way, what benefit does the separation add to LJM users?

The NoiseSuppressor uses WebAudio functionality, when I implemented the voice activity detection which is somewhat similar in structure, there was a specific request to move any 'native' calls to LJM hence all the logic that deals with ScriptProcessorNodes, MediaStreams and so forth was moved there. Arguably we could move the RnnoiseProcessor to LJM and simplify things a big but again it was specifically requested that RnnoiseProcessor be something that can be injected into services that want to use it at a lower level, even tough I doubt that if we decide to add some other noise processor we'll find a common interface.

**Comment by saghul** [comment]
Dammit, I lost the comment I wrote here 2 days ago, oh well, here it goes again.

AFAIS the only existing functionality necessary in LJM is `createAudioContext` which is pretty much a one liner and we can likely get rid of it.

I don't see the point in putting half of the stuff in LJM when it's not in a usable state by LJM users.

I'd rather have the whole effect in JM, just like virtual backgrounds for example. LJM uses who would want to use the effects an lift them from JM and adapt them to their apps.

File: react/features/noise-suppression/index.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (saghul) [inline react/features/noise-suppression/index.js:1]: Don't create an index file for the feature please.

File: conference.js [hunk 2]
```diff
@@ -1733,6 +1735,43 @@ export default {
...
+     * {@code NoiseSuppressionEffect}.
+     */
+    async toggleNoiseSuppression() {
+        const state = APP.store.getState();
+        const dispatch = APP.store.dispatch;
+        const noiseSuppressionActive = isNoiseSuppressionActive(state);
+        const localAudio = getLocalJitsiAudioTrack(APP.store.getState());
+
+        if (!localAudio) {
+            logger.warn('Can not toggle noise suppression without any local track active.');
```
Comment (saghul) [inline conference.js:1749]: Shouldn't this be a notification? Should the control be disabled so we don't end up here?

File: react/features/noise-suppression/actionTypes.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+// @flow
```
Comment (saghul) [inline react/features/noise-suppression/actionTypes.ts:1]: Drop the flow comment.

File: react/features/noise-suppression/actions.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+// @flow
```
Comment (saghul) [inline react/features/noise-suppression/actions.js:1]: General comment, for all new files: use TypeScript please.

File: react/features/noise-suppression/actions.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+import { SET_NOISE_SUPPRESSION_STATE, TOGGLE_NOISE_SUPPRESSION } from './actionTypes';
+
+/**
+ * Updates the noise suppression active state.
+ *
+ * @param {boolean} noiseSuppressionActive - is noise suppression active
+ * @returns {{
+ *      type: SET_NOISE_SUPPRESSION_STATE,
+ *      captureFrameRate: number
```
Comment (saghul) [inline react/features/noise-suppression/actions.js:11]: Bad jsdoc

File: react/features/noise-suppression/actions.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * }}
+ */
+export function setNoiseSuppressionState(noiseSuppressionActive: boolean) {
+    return {
+        type: SET_NOISE_SUPPRESSION_STATE,
+        isNoiseSuppressionActive: noiseSuppressionActive
```
Comment (saghul) [inline react/features/noise-suppression/actions.js:17]: why not call the property "active" ?

File: react/features/noise-suppression/components/NoiseSuppressionButton.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+     * Handles clicking / pressing the button.
+     *
+     * @private
+     * @returns {void}
+     */
+    _handleClick() {
+        const { dispatch } = this.props;
+
+        dispatch(toggleNoiseSuppression());
+        dispatch(setOverflowMenuVisible(false));
```
Comment (saghul) [inline react/features/noise-suppression/components/NoiseSuppressionButton.js:49]: Why do we need to do this?

File: react/features/noise-suppression/middleware.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * @param {Store} store - The redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register(store => next => action => {
+    const result = next(action);
+
+    switch (action.type) {
+    case TOGGLE_NOISE_SUPPRESSION:
+        // Handle noise suppression logic in `conference.js`
+        APP.UI.emitEvent(UIEvents.TOGGLE_NOISE_SUPPRESSION);
```
Comment (saghul) [inline react/features/noise-suppression/middleware.js:22]: Why does this need to be handled in conference.js, if all the operations are performed in the track which is in redux already?

File: react/features/noise-suppression/reducer.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+ * Reduces the Redux actions of the feature features/noise-suppression.
+ */
+ReducerRegistry.register('features/noise-suppression', (state = {}, action) => {
+    const { isNoiseSuppressionActive } = action;
+
+    switch (action.type) {
+    case SET_NOISE_SUPPRESSION_STATE:
+        return {
+            ...state,
+            isNoiseSuppressionActive
```
Comment (saghul) [inline react/features/noise-suppression/reducer.js:18]: i'd just call this `active` the feature name already provides the necessary context.

File: react/features/rtcstats/middleware.js [hunk 10]
```diff
@@ -25,7 +25,6 @@ MiddlewareRegistry.register(store => next => action => {
     const config = state['features/base/config'];
     const { analytics } = config;
 
-
```
Comment (saghul) [inline react/features/rtcstats/middleware.js:28]: Unnecessary change.

File: react/features/stream-effects/rnnoise/RnnoiseProcessor.js [hunk 11]
```diff
@@ -175,23 +133,33 @@ export default class RnnoiseProcessor {
...
-     * @param {Float32Array} pcmFrame - Array containing 32 bit PCM samples.
+     * @param {Float32Array} pcmFrame - Array containing 32 bit PCM samples. Parameter is also used as output
+     * when {@code shouldDenoise} is true.
+     * @param {boolean} shouldDenoise - Should the denoised frame be returned in pcmFrame.
      * @returns {Float} Contains VAD score in the interval 0 - 1 i.e. 0.90 .
      */
-    calculateAudioFrameVAD(pcmFrame: Float32Array) {
-        if (this._destroyed) {
-            throw new Error('RnnoiseProcessor instance is destroyed, please create another one!');
+    calculateAudioFrameVAD(pcmFrame: Float32Array, shouldDenoise: Boolean = false) {
```
Comment (saghul) [inline react/features/stream-effects/rnnoise/RnnoiseProcessor.js:141]: Can we detach the functionality here? I find it very confusing that an API described as "Calculate the Voice Activity Detection for a raw Float32 PCM sample Array." also denoises. the audio.

File: react/features/stream-effects/rnnoise/RnnoiseProcessor.js [hunk 12]
```diff
@@ -175,23 +133,33 @@ export default class RnnoiseProcessor {
...
+        // Convert 32 bit Float PCM samples to 16 bit Float PCM samples as that's what rnnoise accepts as input
+        for (let i = 0; i < 480; i++) {
```
Comment (saghul) [inline react/features/stream-effects/rnnoise/RnnoiseProcessor.js:143]: Pl add a constant for the samples amount with some docs.

File: react/features/stream-effects/rnnoise/RnnoiseProcessor.js [hunk 13]
```diff
@@ -175,23 +133,33 @@ export default class RnnoiseProcessor {
...
+            this._wasmInterface.HEAPF32[this._wasmPcmInputF32Index + i] = pcmFrame[i] * 32768;
```
Comment (saghul) [inline react/features/stream-effects/rnnoise/RnnoiseProcessor.js:144]: Please add a constant for the magic number.

File: react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        return sourceLocalTrack.isAudioTrack();
+    }
+
+    /**
+     * Clean up resources acquired by noise suppressor and rnnoise processor.
+     *
+     * @returns {void}
+     */
+    stopEffect() {
+        this._noiseSuppressor && this._noiseSuppressor.destroy();
```
Comment (saghul) [inline react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.js:64]:
```suggestion
        this._noiseSuppressor?.destroy();
```

File: react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        this._rnnoiseProcessor && this._rnnoiseProcessor.destroy();
```
Comment (saghul) [inline react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.js:65]:
```suggestion
        this._rnnoiseProcessor?.destroy();
```

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments PTAL! Some additional general questions:

- should we disable WebRTC's builtin NS when using this?
- given how sensitive 8x8 is to audio quality, shouldn't we be running this in an Audio Worklet to make sure an overloaded main thread doesn't compromise audio quality?

**Comment by andrei-gavrilescu** [comment]
> Dammit, I lost the comment I wrote here 2 days ago, oh well, here it goes again.
> 
> AFAIS the only existing functionality necessary in LJM is `createAudioContext` which is pretty much a one liner and we can likely get rid of it.
> 
> I don't see the point in putting half of the stuff in LJM when it's not in a usable state by LJM users.
> 
> I'd rather have the whole effect in JM, just like virtual backgrounds for example. LJM uses who would want to use the effects an lift them from JM and adapt them to their apps.

Well the NoiseSuppressor is usable without JM as long as you provide a dennoiseProcessor which follows the expected interface, doesn't have to be rnnoise (even though in reality it will probably be rnnoise). 
Also regarding createAudioContext, that's just the create, but then we have the calls around it like `createMediaStreamSource` `createScriptProcessor` etc. From my understanding, those would have to be abstracted as well so we don't have "native" calls in JM. And if that's the case is it worth creating such an abstraction for something we'll remove as soon as we get a chance?

**Comment by saghul** [comment]
> Well the NoiseSuppressor is usable without JM as long as you provide a dennoiseProcessor which follows the expected interface, doesn't have to be rnnoise (even though in reality it will probably be rnnoise).

I generally push back in adding generic APIs when we only have 1 single implementation. We don't know if it will work for other desoisers. Removing existing APIs is hard, so I'd rather limit the scope and open it up later if need be.



> Also regarding createAudioContext, that's just the create, but then we have the calls around it like `createMediaStreamSource` `createScriptProcessor` etc. From my understanding, those would have to be abstracted as well so we don't have "native" calls in JM.

I don't see why that is necessary. This feature is web only, so there is nothing to abstract for native, the code will not run there. There is no way for you to get audio frames in JS on native, and there likely never will due to RN's architecture.

File: react/features/noise-suppression/components/NoiseSuppressionButton.js (hunk 7 above)
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/components/NoiseSuppressionButton.js:49]: in order to hide the `more` menu after the button is clicked

File: react/features/noise-suppression/index.js (hunk 1 above)
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/index.js]: removed

**Comment by andrei-gavrilescu** [comment]
> * should we disable WebRTC's builtin NS when using this?

I've tested it with the WebRTC NS on and off, it doesn't seem to make much of a difference.

File: react/features/stream-effects/rnnoise/RnnoiseProcessor.js (hunk 13 above)
Comment (andrei-gavrilescu) [reply react/features/stream-effects/rnnoise/RnnoiseProcessor.js]: added

File: react/features/stream-effects/rnnoise/RnnoiseProcessor.js (hunk 12 above)
Comment (andrei-gavrilescu) [reply react/features/stream-effects/rnnoise/RnnoiseProcessor.js]: added

File: react/features/stream-effects/rnnoise/RnnoiseProcessor.js (hunk 11 above)
Comment (andrei-gavrilescu) [reply react/features/stream-effects/rnnoise/RnnoiseProcessor.js]: I've renamed the function to something a bit more suggestive

File: react/features/noise-suppression/actions.js (hunk 6 above)
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/actions.js]: changed to active

File: react/features/noise-suppression/actions.js (hunk 5 above)
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/actions.js]: fixed

File: react/features/noise-suppression/actionTypes.ts (hunk 3 above)
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/actionTypes.ts]: dropped

File: conference.js (hunk 2 above)
Comment (andrei-gavrilescu) [reply conference.js]: added notifications

File: react/features/stream-effects/rnnoise/RnnoiseProcessor.ts [hunk 16]
```diff
@@ -66,22 +61,13 @@ export default class RnnoiseProcessor {
             // For VAD score purposes only allocate the buffers once and reuse them
             this._wasmPcmInput = this._wasmInterface._malloc(RNNOISE_BUFFER_SIZE);
 
+            // eslint-disable-next-line no-bitwise
```
Comment (saghul) [inline react/features/stream-effects/rnnoise/RnnoiseProcessor.ts:64]: Feel free to disable it for the whole file.

File: webpack.config.js [hunk 17]
```diff
@@ -392,6 +392,39 @@ module.exports = (_env, argv) => {
                 ...getBundleAnalyzerPlugin(analyzeBundle, 'face-landmarks-worker')
             ],
             performance: getPerformanceHints(perfHintOptions, 1024 * 1024 * 2)
+        }),
+        Object.assign({}, config, {
+            /**
+             * The NoiseSuppressorWorklet is loaded in an audio worklet which doesn't have the same
```
Comment (saghul) [inline webpack.config.js:398]: 😱

File: webpack.config.js [hunk 18]
```diff
@@ -418,7 +418,7 @@ module.exports = (_env, argv) => {
             ] },
             plugins: [
             ],
-            performance: getPerformanceHints(perfHintOptions, 40 * 1024),
+            performance: getPerformanceHints(perfHintOptions, 1024 * 1024),
```
Comment (saghul) [inline webpack.config.js:421]: Doesn't that look very big? Or is it because of rnnoise?

File: react/features/stream-effects/rnnoise/RnnoiseProcessor.ts (hunk 16 above)
Comment (andrei-gavrilescu) [reply react/features/stream-effects/rnnoise/RnnoiseProcessor.ts:64]: done

File: react/features/stream-effects/rnnoise/RnnoiseProcessor.ts [hunk 19]
```diff
@@ -15,6 +13,8 @@ const RNNOISE_BUFFER_SIZE: number = RNNOISE_SAMPLE_LENGTH * 4;
  */
 const PCM_FREQUENCY: number = 44100;
 
+const SHIFT_16_BIT_NR: number = 32768;
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/stream-effects/rnnoise/RnnoiseProcessor.ts:16]: added comments

File: react/features/stream-effects/noise-suppression/NoiseSuppressorWorklet.ts [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+        }
+
+        // Because the circular buffer's length is the lcm of both input size and the processor's sample size,
+        // by the time we reach the end with the input index the denoise length index will be there as well.
+        if (this._inputBufferLength === this._circularBufferLength) {
+            this._inputBufferLength = 0;
+            this._denoisedBufferLength = 0;
+        }
+
+        return true;
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/stream-effects/noise-suppression/NoiseSuppressorWorklet.ts:167]: Yes, they need to return true in order for the next iteration to occur. The flow from [here](https://github.com/jitsi/jitsi-meet/blob/e5572c556ad881a4a81f6cbfc86f38e57e950d1b/react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.ts#L73) should technically close them, however there seems to be a bug around this on chrome, check the comments from link.

File: react/features/noise-suppression/middleware.js [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * @param {Store} store - The redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register(({ dispatch, getState }) => next => async action => {
+    const result = next(action);
+
+    switch (action.type) {
+    case TOGGLE_NOISE_SUPPRESSION: {
+
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/middleware.js:22]: removed

[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/middleware.js:22]: removed

File: react/features/noise-suppression/middleware.js [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        if (!localAudio) {
+            logger.warn('Can not toggle noise suppression without any local track active.');
+
+            dispatch(showWarningNotification({
+                titleKey: 'notify.noiseSuppressionFailedTitle',
+                descriptionKey: 'notify.noiseSuppressionNoTrackDescription'
+            }, NOTIFICATION_TIMEOUT_TYPE.MEDIUM));
+
+            return;
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/middleware.js:35]: added

File: react/features/noise-suppression/components/NoiseSuppressionButton.js [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Maps part of the Redux state to the props of this component.
+ *
+ * @param {Object} state - The Redux state.
+ * @private
+ * @returns {Props}
+ */
+function _mapStateToProps(state: Object): $Shape<Props> {
+
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/components/NoiseSuppressionButton.js:72]: removed

File: react/features/stream-effects/noise-suppression/NoiseSuppressorWorklet.ts [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+     * @param {Float32Array[][]} inputs - Array of inputs connected to the node, each of them with their associated
+     * array of channels. Each channel is an array of 128 pcm samples.
+     * @param {Float32Array[][]} outputs - Array of outputs similar to the inputs parameter structure, expected to be
+     * filled during the execution of `process`. By default each channel is zero filled.
+     * @returns {boolean} - Boolean value that returns whether or not the processor should remain active. Returning
+     * false will terminate it.
+     */
+    process(inputs: Float32Array[][], outputs: Float32Array[][]) {
+
+         // We expect the incoming track to be mono, if a stereo track is passed only on of its channels will get
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/stream-effects/noise-suppression/NoiseSuppressorWorklet.ts:102]: added check before starting the worklet.

File: react/features/noise-suppression/logger.js [hunk 25]
```diff
@@ -0,0 +1,5 @@
+// @flow
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/logger.js]: 🔥

File: react/features/noise-suppression/components/NoiseSuppressionButton.js [hunk 26]
```diff
@@ -0,0 +1,78 @@
+// @flow
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/components/NoiseSuppressionButton.js]: 🔥ed

File: react/features/noise-suppression/actionTypes.ts [hunk 27]
```diff
@@ -0,0 +1,11 @@
...
+ * Type of action which sets the current state of noise suppression.
+ *
+ * {
+ *     type: SET_NOISE_SUPPRESSION_STATE,
+ *     active: boolean
+ * }
+ */
+export const SET_NOISE_SUPPRESSION_STATE = 'SET_NOISE_SUPPRESSION_STATE';
+
+export const TOGGLE_NOISE_SUPPRESSION = 'TOGGLE_NOISE_SUPPRESSION';
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/actionTypes.ts:18]: added

File: react/features/noise-suppression/middleware.ts [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+
+        if (!localAudio) {
+            logger.warn('Can not toggle noise suppression without any local track active.');
+
+            dispatch(showWarningNotification({
+                titleKey: 'notify.noiseSuppressionFailedTitle',
+                descriptionKey: 'notify.noiseSuppressionNoTrackDescription'
+            }, NOTIFICATION_TIMEOUT_TYPE.MEDIUM));
+
+            return false;
```
Comment (saghul) [inline react/features/noise-suppression/middleware.ts:77]: You need to return `result` here or the middleware chain will enter some inconsistent state.

File: react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.ts [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+        this._audioContext = new AudioContext();
+
+        this._audioSource = this._audioContext.createMediaStreamSource(audioStream);
+        this._audioDestination = this._audioContext.createMediaStreamDestination();
+
+        const baseUrl = `${getBaseUrl()}libs/`;
+        const workletUrl = `${baseUrl}noise-suppressor-worklet.min.js`;
+
+        // Connect the audio processing graph MediaStream -> AudioWorkletNode -> MediaStreamAudioDestinationNode
+        this._audioContext.audioWorklet.addModule(workletUrl).then(() => {
```
Comment (saghul) [inline react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.ts:48]: What if this fails? I feel like we need a `catch` here, and potentially some logging / handling of the problem.

File: tsconfig.json [hunk 30]
```diff
@@ -9,7 +9,8 @@
         "noEmit": false,
         "moduleResolution": "Node",
         "strict": true,
-        "noImplicitAny": true
+        "noImplicitAny": true,
+        "strictPropertyInitialization": false
```
Comment (saghul) [inline tsconfig.json:13]: Why did you add this?

**Review by saghul — COMMENTED** [review]
Left some comments, PTAL. I only glanced through the math, I assume you got that right :-)

File: react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.ts (hunk 29 above)
Comment (andrei-gavrilescu) [reply react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.ts:48]: startEffect is called by the JitsiLocalTrack and any potential exceptions should be bubbled up and handled [here ](https://github.com/jitsi/lib-jitsi-meet/blob/634885b9fb513103a690b43396d8369544b50cf3/modules/RTC/JitsiLocalTrack.js#L881,) or at least that's my understanding of it.
Idk if there's anything actionable at this point in case of an exception, so I just let the JitsiLocalTrack handle the logging.

Comment (saghul) [reply react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.ts:48]: But the failure may happen later, since it's a promise chain.

File: react/features/noise-suppression/actions.ts [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+        type: TOGGLE_NOISE_SUPPRESSION
+    };
+}
+
+/**
+ * Turns off noise suppression if it's already enabled
+ *
+ * @returns {Function}
+ */
+ export function turnOffNoiseSuppression() : Function {
```
[the remark this answers is not in the document]
Comment (saghul) [reply react/features/noise-suppression/actions.ts:38]: Feels a bit convoluted, no? I'd say TOGGLE_NOISE_SUPPRESSION should not be the action we act on, but a helper that reads the local state and dispatches the opposite. Then we can simplify this by removing this extra action and acting on SET_NOISE_SUPPRESSION_STATE directly.

Do you see a problem with that approach? I might be missing something.

[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/actions.ts:38]: I guess, this might stem from my lacking knowledge of redux/middleware/react. So the reason why I have two of them was so that in case of an error other interested parties (middlewares) wouldn't act on SET_NOISE_SUPPRESSION_STATE, whereas if it's just the one, other middleware's wouldn't know if the outcome was successful or failed and still act on it. But I'm assuming there are ways of doing this with a single action? like preventing it from reaching the reducer in case of an error, not sure how though.

File: tsconfig.json (hunk 30 above)
Comment (andrei-gavrilescu) [reply tsconfig.json:13]: without it typescript forces me to initialise all properties in the c'tor.

File: react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.ts (hunk 29 above)
Comment (andrei-gavrilescu) [reply react/features/stream-effects/noise-suppression/NoiseSuppressionEffect.ts]: I see, will add a catch with a log there.

File: react/features/noise-suppression/middleware.ts (hunk 28 above)
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/middleware.ts]: woops, added.

File: react/features/noise-suppression/functions.ts [hunk 32]
```diff
@@ -0,0 +1,11 @@
+import { IState } from "../app/types";
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/functions.ts]: rebased, should be ok now.

File: react/features/noise-suppression/actions.ts [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+    return async (dispatch: Function, getState: Function) => {
+        if (isNoiseSuppressionActive(getState())) {
+            await dispatch(toggleNoiseSuppression());
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/actions.ts]: removed.

File: tsconfig.json (hunk 30 above)
Comment (saghul) [reply tsconfig.json:13]: That sounds like A Good Thing though ;-)

File: react/features/noise-suppression/actions.ts (hunk 31 above)
[the remark this answers is not in the document]
Comment (saghul) [reply react/features/noise-suppression/actions.ts]: Let's sync on this.

File: tsconfig.json (hunk 30 above)
Comment (andrei-gavrilescu) [reply tsconfig.json:13]: I mean, the typescript docs don't seem to advise against them. And not having to type the name again in the c'tor seemed like a plus to me.
But if you feel strongly about it I'll remove and change the classes.

Comment (saghul) [reply tsconfig.json:13]: No strong feeling, let's try it out.

File: react/features/noise-suppression/actions.ts (hunk 33 above)
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply react/features/noise-suppression/actions.ts]: I've added this back, it seems that even though it's not a thunk that's being returned, dispatch will return a promise which awaits for the completion of the async middleware (I think). In this case it awaits for the NS middleware to finish the setEffect(undefined) action. We need to wait for it to finish otherwise any subsequent effects set on that track might throw a 'setEffect change already in progress error'.
The case I'm trying to address here is :
1.  User has noise suppression activated.
2. User start audio screen sharing.
3. Audio screen sharing implies adding a different effect to the current local track, because with the current implementation we can't have multiple effects on a single track we need to remove the ns effect, this is done via the turnOffNoiseSuppression that will remove it in the associated middleware and update the state accordingly. Basically we need to wait for the async middleware called by the dispatch to finish, and it seems await does the trick.
An alternative would be to not allow the user to start audio sharing while NS is active, but this flow felt more natural. Even if we were to use the alternative flow, the await here seems to have an effect, do you know why? or is there a different approach we could take if this feels hacky?

**Comment by andrei-gavrilescu** [comment]
Jenkins please test this please.

**Review by saghul — APPROVED** [review]
Impressive work! 🚀
